spacy/tokens/doc.pyx

```diff
@@ -1743,6 +1743,124 @@ cdef class Doc:
                     j += 1
         return output
 
+    @cython.boundscheck(False)  # Deactivate bounds checking
+    def get_character_combination_hashes(self,
```

Comment:
Does this need to be added to `Doc`? This seems functionality that is relatively tightly coupled to the rich feature extractor and `Doc.c` is accessible to Cython code outside `Doc`.

Reply:
I think `Doc` is probably the wrong place, but wasn't sure where else to put it: do you have any suggestions?

Reply:
Can't it stay in the `richfeatureextrator` module (after converting it to a Cython module)?

Reply:
@adrianeboyd @svlandeg what do you think of @shadeMe's suggestion? Would putting everything in `richfeatureextractor` conform to all explicit and implicit standards around where things go within spaCy?

Reply:
I can definitely see Daniël's & Madeesh's reasoning to keep this closer to its related functionality - i.e. in the `richfeatureextractor`.

Reply:
I tried moving all the functionality to `richfeatureextractor`, but came across the problem that the `get_character_combination_hashes()` method directly accesses the C/Cython member variables of the `Doc` object it is analyzing, specifically `TokenC* c` and `StringStore vocab.strings`, and needs to do so for speed. Is there a way of accessing these from another Cython module that has been called from Python with a `Doc` object as a parameter? There may well be one I am not aware of as I am far from a Cython expert! If there is none, though, it looks as though the `get_character_combination_hashes()` method may need to stay within `Doc`, although there would obviously be scope for moving the private methods like `_write_hashes()` into a separate module (probably not `richfeatureextractor` in that case though, because that would lead to a dependency loop).